### PR TITLE
ChainWriter EVM GetFeeComponents implementation

### PR DESCRIPTION
**Pull request by silaslenihan**
See [ChainWriter doc](https://docs.google.com/document/d/1fTJU1sEwnmWsejWzYtTdtv8uXy1pD4eM9d3Wy4wsAxo/edit#heading=h.rtroriuud1ph) for context
Ticket: [BCI-3367](https://smartcontract-it.atlassian.net/jira/software/c/projects/BCI/boards/23?assignee=61a6f090ef18ca00717f84db&quickFilter=398&selectedIssue=BCI-3367)

[BCI-3367]: https://smartcontract-it.atlassian.net/browse/BCI-3367?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/services/relay/evm/chain_writer.go
Comment (patrick-dowell) [inline core/services/relay/evm/chain_writer.go:195]: I wonder what kind of errors are possible here. Are there situations where we'd want to return both the executionFee and the l1Oracle RPC error?

File: core/services/relay/evm/chain_writer.go
Comment (patrick-dowell) [inline core/services/relay/evm/chain_writer.go]: I know this code here is copied from somewhere else, but I think it'd be helpful to describe what's going on here in the comments. In particular, what's the difference between the DynamicFeeCap and the Legacy Gas Price? What's the situation in which we fall back to legacy, ie why would the DynamicFeeCap be nil?

File: core/services/relay/evm/chain_writer.go
Comment (patrick-dowell) [inline core/services/relay/evm/chain_writer.go]: do we have what maxGasPrice is and how it's used documented?

**Review by patrick-dowell — COMMENTED** [review]
Overall looks reasonable. I know that we want to split out the unit tests into a separate PR, so just left comments on the ChainWriter logic.

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Added comment explaining this above the method signature

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: `maxGasPrice` is used by the Evm Estimator as an override of the `feeCap` for the transaction. I added a comment in the struct declaration explaining this

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go:195]: Hm, are you suggesting we don't return an error if the gasPrice errors out and wait to see if the L1Oracle is erroring out too, returning both errors in that case?

Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go:195]: Hm, I'm not sure actually. I think we would want to return nil and an error because the L1Oracle fee is required for estimating gas. I suppose it doesn't hurt to return the execution fee though

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: Why use a `%+v` here? That seems unusual

File: core/services/relay/evm/types/types.go
Comment (jmank88) [inline core/services/relay/evm/types/types.go]: Why do we need a new field for this, given that we already have the ability to override per chain and per account?

Comment (jmank88) [reply core/services/relay/evm/types/types.go]: Actually, I'm confused about where this is sourced from. I see that this is not in the TOML, but I don't see it being set from anything else. How is this expected to be used?

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: Why are these packaged together? Would it be simpler to just have two field on the chainWriter?

File: core/services/relay/evm/types/types.go
Comment (jmank88) [reply core/services/relay/evm/types/types.go]: The existing config API consists of a func for looking up by account dynamically. Doesn't that need to be supported? https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/config/chain_scoped_gas_estimator.go#L17

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: This was copied directly from CCIP's [gas estimation](https://github.com/smartcontractkit/ccip/blob/4a6f68e052f1e55596b1f8238f23806a7128c541/core/services/ocr2/plugins/ccip/prices/exec_price_estimator.go#L39), since they are the main consumer of this function. I suppose they wanted to include the names of the fields in the error message.

File: core/services/relay/evm/types/types.go
Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: Again, this functionality mimics the [CCIP gas esimtation](https://github.com/smartcontractkit/ccip/blob/4a6f68e052f1e55596b1f8238f23806a7128c541/core/services/ocr2/plugins/ccip/prices/exec_price_estimator.go#L23). I suppose we could remove it entirely and rely on the gas estimator's config instead, is that what you are suggesting?

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Good point, I'll remove the `CWGasEstimator` struct

**Comment by jmank88** [comment]
Is there a ticket to link in the description?

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [reply core/services/relay/evm/chain_writer.go]: Does it make sense to you why the internals of the type should be exposed? It does not make sense to me and it inherently carries risk.

None of the field in this type are even called 'price':
```go
type EvmFee struct {
	// legacy fees
	Legacy *assets.Wei

	// dynamic/EIP1559 fees
	DynamicFeeCap *assets.Wei
	DynamicTipCap *assets.Wei
}
```

IMHO this error should be telling the user what happened instead of making them figure it for themselves. `gasPrice == nil` means that we failed to derive the gas price from this other type, but the reason is that both `Legacy` and `DynamicFeeCap` were nil, so that is what the error should highlight.

File: core/services/relay/evm/types/types.go
Comment (jmank88) [reply core/services/relay/evm/types/types.go]: I don't understand well enough what you are trying to achieve to say what should be happening. Do you intend to use this variable anywhere at this time? I do not see it being set anywhere in this PR.

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: core/services/relay/evm/types/types.go
Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: @jmank88 Sorry - for some reason I didn't see the latter two comments when I initially responded. I'm not sure I understand your question. When a product that is using the ChainWriter needs the ChainWriter to be capable of gas estimation, it will include a `MaxGasPrice` in the config that is passed into `NewChainWriterService`. I'm not very familiar with the `PriceMaxKey` function you shared but I supposed `MaxGasPrice` in the ChainWriter would be used to override that value in the gas estimator's `GetFee` function. For example the EVM gas estimators take the [min between the user defined max and the one specified in the gas estimators config. ](https://github.com/smartcontractkit/chainlink/blob/4fdfffdb4c4a6f16fcdb388fe38642398ce130da/core/chains/evm/gas/models.go#L510) So here, we are just enabling that override functionality. Does that make sense or am I failing to grasp something here?

Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: The reason it isn't being set anywhere in this PR is because the [Keystone Write Target](https://github.com/smartcontractkit/chainlink/blob/f0d2ca52ead9ba3136bafc8e2c5cf7915f858ee3/core/services/relay/evm/write_target.go#L57-L73) doesn't need gas estimation. When CCIP implements the ChainWriter they will have to configure `MaxGasPrice` in the `ChainWriterConfig` to enable gas esitmation.

Comment (jmank88) [reply core/services/relay/evm/types/types.go]: The max gas price is currently something that is configured per account, and thus completely out of the control of the sender of a transaction. When you send through a particular account, you are implicitly opting in to that account's maximum gas price. I don't understand how what we are doing here will connect to the existing system, and if we don't actually have a use for this code right now, why are we building it?

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: The linter should be flagging these as unsorted :thinking:

File: core/services/relay/evm/types/types.go
Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: > The max gas price is currently something that is configured per account, and thus completely out of the control of the sender of a transaction. 

I don't understand this statement, since [GetFee](https://github.com/smartcontractkit/chainlink/blob/4fdfffdb4c4a6f16fcdb388fe38642398ce130da/core/chains/evm/gas/models.go#L38) accepts a `maxFeePrice`. This PR isn't changing that logic in any way.

> and if we don't actually have a use for this code right now, why are we building it

Because this functionality is required for CCIPv2 to use the `ChainWriter`

Comment (jmank88) [reply core/services/relay/evm/types/types.go]: > I don't understand this statement, since [GetFee](https://github.com/smartcontractkit/chainlink/blob/4fdfffdb4c4a6f16fcdb388fe38642398ce130da/core/chains/evm/gas/models.go#L38) accepts a maxFeePrice. This PR isn't changing that logic in any way.

Here `GetFee` is called with a `keySpecificGasPrice` from tx attempt builder:
https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/txmgr/attempts.go#L60-L61

Comment (jmank88) [reply core/services/relay/evm/types/types.go]: > Because this functionality is required for CCIPv2 to use the ChainWriter

Do you mean that there will be a usage in the CCIP repo? please link any corresponding PRs

Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: > Here GetFee is called with a keySpecificGasPrice from tx attempt builder:
https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/txmgr/attempts.go#L60-L61

Ahh I see, thanks for clarifying. I wonder what address we could use here since `GetFeeComponents` is used outside of the context of a specific Tx / address. Maybe we need to require the caller to pass an address? 

> Do you mean that there will be a usage in the CCIP repo? please link any corresponding PRs

There are no PRs yet in the CCIP repo to my knowledge, I believe they are waiting for us to complete the EVM implementation of the ChainWriter.

File: core/services/relay/evm/chain_writer.go
Comment (dimriou) [inline core/services/relay/evm/chain_writer.go]: I understand this was copy pasted, but I find the usage of `gasPriceWei` quite confusing since what is returned is not a gas price value but rather an `EvmFee` object instead.

File: core/services/relay/evm/chain_writer.go
Comment (nickcorin) [reply core/services/relay/evm/chain_writer.go:195]: The usual practice for returning a tuple that follows the `(value, error)` is for one and only one to be non-nil. I'd return `nil` here.

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Good point, I renamed this object `fee`

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Updated the error message to be more specific

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]:
```suggestion
// GetFeeComponents the execution and data availability (L1Oracle) fees for the chain.
```

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Are these not sorted?

Comment (jmank88) [reply core/services/relay/evm/chain_writer.go]: No, the whole import section should be ordered:
1. std lib
2. third party
3. smartcontractkit
4. this module

It is OK to have multiple groupings within 2 & 3, but packages from the same module should always stay grouped together.

Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Huh, that's strange. When I run `golangci-lint run -v --fix ./core/services/relay/evm` I get no warnings or anything and obviously the CI linting test is catching this. When you lint locally does it fix / warn about this?

File: core/services/relay/evm/chain_writer.go
Comment (patrick-dowell) [inline core/services/relay/evm/chain_writer.go:45]: would it make sense to have a default value?

File: core/services/relay/evm/write_target.go
Comment (patrick-dowell) [inline core/services/relay/evm/write_target.go]: shouldn't we set the gas estimator properly here?

**Review by patrick-dowell — COMMENTED** [review]
Two small suggestions but otherwise looks good to me.

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go:42]: I don't think that would be necessary since this is already a required config in the gas estimator https://github.com/smartcontractkit/chainlink/blob/4fdfffdb4c4a6f16fcdb388fe38642398ce130da/core/chains/evm/config/config.go#L134

File: core/services/relay/evm/write_target.go
Comment (silaslenihan) [reply core/services/relay/evm/write_target.go]: I didn't think that was necessary since Keystone doesn't use `GetFeeComponents` yet.

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [reply core/services/relay/evm/chain_writer.go]: The tools as configured will only separate std lib, third party, and local module. Keeping things organized beyond that is up to us.

Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Ah ok, I just reordered them manually

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go:45]: Added a default of 500 GWEI

**Review by patrick-dowell — DISMISSED** [review]
I'm good at this point. Once we get Jordan's approval, we can merge.

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: What is this value based on? We already have mechanisms for configuration defaults and overrides. I don't think it is a good idea to splinter configuration like this. It would be simple and clear to return an error if the value is unspecified.

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: This error should probably say something different now

File: core/services/relay/evm/chain_writer.go
Comment (jmank88) [inline core/services/relay/evm/chain_writer.go]: Error messages are meant to be lower case:
```suggestion
		return nil, fmt.Errorf("dynamic fee and legacy gas price missing %+v", fee)
```
Linter is supposed to catch this one... :thinking:

File: core/services/relay/evm/write_target.go
Comment (jmank88) [reply core/services/relay/evm/write_target.go]: I don't understand why we would leave this out. It is trivial to set both the max gas price and estimator since they are already both available in-scope.
```suggestion
	chainWriterConfig.MaxGasPrice = chain.Config().EVM().GasEstimator().PriceMax()
	cw, err := NewChainWriterService(lggr.Named("ChainWriter"), chain.Client(), chain.TxManager(), chain.GasEstimator(), chainWriterConfig)
```

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: changed error to "gas estimator not available"

File: core/services/relay/evm/chain_writer.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_writer.go]: Updated to return an error

File: core/services/relay/evm/types/types.go
Comment (jmank88) [inline core/services/relay/evm/types/types.go]: Why don't we use `assets.Wei` here instead of converting back and forth?

Comment (silaslenihan) [reply core/services/relay/evm/types/types.go]: Good point, just fixed this.

**Review by jmank88 — DISMISSED** [review]

**Review by patrick-dowell — DISMISSED** [review]
LGTM

**Review by nickcorin — DISMISSED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by patrick-dowell — APPROVED** [review]

**Pull request merged**
